task_manager/task_manager.hh

```diff
@@ -0,0 +1,323 @@
+                task->_children = std::move(tasks);
+                return make_ready_future<std::vector<task_id>>(boost::copy_range<std::vector<task_id>>(task->_children
+                    | boost::adaptors::transformed([] (foreign_task_ptr& child) { return child->_status.id; })));
+            });
+        }
+
+        future<> done() const noexcept {
+            return _done.get_shared_future();
+        }
+
+        void unregister() {
+            _module->unregister_task(_status.id);
+        }
+
+        void finish() noexcept {
+            _status.end_time = db_clock::now();
+            _status.state = task_manager::task_state::done;
+            _done.set_value();
+        }
+
+        void finish_failed(std::exception_ptr ex, std::string error) noexcept {
+            _status.end_time = db_clock::now();
+            _status.state = task_manager::task_state::failed;
+            _status.error = std::move(error);
+            _done.set_exception(ex);
+        }
+
+        void finish_failed(std::exception_ptr ex) noexcept {
+            finish_failed(ex, fmt::format("{}", ex));
+        }
+
+        virtual void abort() = 0;
+    };
+    class module {
+    private:
+        task_manager& _tm;
+        std::string _name;
+        task_map _tasks;
+        seastar::gate _gate;
+    public:
+        module(task_manager& tm, std::string name) : _tm(tm), _name(std::move(name)) {} // TODO Deexie protect from global usages.
+
+        static module_ptr make_module(task_manager& tm, std::string name) {
+            auto m = seastar::make_shared<module>(tm, name);
+            tm.register_module(std::move(name), m);
+            return m;
+        }
+
+        seastar::gate& async_gate() noexcept {
+            return _gate;
+        }
+
+        abort_source& abort_source() noexcept {
+            return _tm.abort_source();
+        }
+
+        std::chrono::seconds get_task_ttl() const {
+            return _tm.get_task_ttl();
+        }
+
+        task_manager::task_map& get_tasks() noexcept {
+            return _tasks;
+        }
+
+        const task_manager::task_map& get_tasks() const noexcept {
+            return _tasks;
+        }
+
+        task_manager::task_map& get_all_tasks() noexcept {
+            return _tm.get_all_tasks();
+        }
+
+        const task_manager::task_map& get_all_tasks() const noexcept {
+            return _tm.get_all_tasks();
+        }
+
+        void register_task(task_ptr task) {
+            _tasks[task->get_status().id] = task;
+            _tm.register_task(task);
+        }
+
+        void unregister_task(task_id id) {
+            if (_tasks.contains(id)) {
+                _tasks.erase(id);
+                _tm.unregister_task(id);
+            }
+        }
+
+        future<> stop() noexcept {
+            return _gate.close().then([this] {
+                _tm.unregister_module(_name);
+            });
+        }
+    private:
+        future<unsigned> get_task_shard(task_id id) const;
```

Comment:
Other than `run_on_shards` (which is not used in this patch), the other private methods aren't implemented in this patch.
Let's introduce them along with their uses in the next patches.

Reply:
Oh, I left them mistakenly trying to provide that all `task_manager` would be available through `task_manager::module`.

What should be done with corresponding methods in `task_manager`? They are used in api implementation. Should they be moved to that patch?

Reply:
Probably.  And it's ok to have some meaningful logic in the api implementation that uses basic primitives that the task_manager classes provide.  E.g. we necessarily need to provide a methods that looks up a module in the modules map using a string.  We can just provide a method that exposes the modules map (as const&) and let the caller at the api layer use it directly.